### ISSUE TITLE
Pattern: Launchy: An Accessible Modal Window

### ISSUE THREAD

**Issue by svinkle** [a11yproject/a11yproject.com#497]
Hi everyone!

I'd like to add this pattern for an accessible modal window and am looking for code review/feedback/approval when you have a free moment.

https://codepen.io/svinkle/pen/pjVepz

Thanks, friends! 🙂

/cc @davatron5000 @joe-watkins @mpiotrowicz @jeryj

**Comment by jeryj** [comment]
Hey @svinkle!

It's looking good! I want to make sure I'm reviewing it in the right context, so I broke my quick feedback down into two sections. 

It looks like this is your goal for the modal pattern: Provide an educational example modal pattern that meets all of the criteria at https://www.w3.org/TR/wai-aria-practices-1.1/#dialog_modal 

If that's the case, it's looking really good. I like how you went about building out the last focused element and focus trap part. The only question I have is why you're using `<a>` elements instead of `<button>`s for opening/closing the modal. 

If your intention is to make it something people will drop into their own production codebase, I'd look at adding:
- support for multiple modals by wrapping into some kind of factory function
- Example CSS to handle different page heights (showing all of the modal if the modal is larger than the window, locking the scrollable area to where the modal is)
- a system for building the modals when JS loads

Cheers!
Jerry

**Comment by joe-watkins** [comment]
@svinkle This is slick! I'd use this in production if there was an option to allow for focus to leave the modal to the browser address bar for non-critical modal window scenarios. Perhaps integration of optional `inert` w/polyfill? https://github.com/WICG/inert That may be overdoing it but might be fun to fiddle around with. :)

I agree w/ @jeryj .. bundle this thing up for mass use.. it is a great pattern and demonstrates many best practices. Maybe create a repo and we can all pound on it :)

p.s. IE 11 is choking on Flexbox/centering

**Comment by svinkle** [comment]
Thanks, @jeryj!

My thoughts exactly. It currently doesn't support multiple modals or handle heights well nor does it progressively enhance. I thought maybe this would be one way to showcase the "computed" HTML? I really should address these concerns though, as you're correct, this isn't copy and paste-able.

As for the old cage match of `<a>` vs `<button>`, I'm coming from the argument that when opening the modal window, focus shifts from one point in the page to another, much like a jump-link would behave. This is the logic and thinking behind using links for launching modal windows.

Let me know if you have other thought on this.
Thanks for looking!

**Comment by jeryj** [comment]
@svinkle I figured you had a really good reason for using `<a>` instead of a `<button>`. The more I thought about it the `<a>` made sense because, if JS wasn't available, you could show the modal in the footer by default and have the close button (er... link 😀 ) return you to the original modal open link. I really liked that idea, but when I tested it w/o JS I couldn't access the modal content. So, if that is the intention, then that's awesome! Thanks for clarifying!

**Comment by svinkle** [comment]
@joe-watkins thanks! Same with Safari, doesn't like the flexbox. I'm kind of making the assumption that people will be more interested in the JS than the CSS, or I'm just being lazy. I should fix it. ☺️

Not really a fan of dropping out of the modal window to get to the browser chrome. Eh, again me being lazy.

I think I will make this its own repo for folks to contribute to. 👍

I figure this would be a decent pattern to use until `inert` becomes a thing. 😁

**Comment by svinkle** [comment]
@jeryj, ideally the content will live in it's proper place in the DOM in a nojs scenario. When this thing is properly PE'd, all there will be is probably a `data-*` attribute of sorts to mark it as a modal window. From the JS can dynamically add all the things it needs to run.

Now I'm pumped to work on this more, lol.

**Comment by svinkle** [comment]
Calling for another look at this. I've updated the script to be fully dynamic and handle more than one instance.

I've moved the JS source from the CodePen over to my GitHub repo as there's a webpack build process now and all that jazz.

Anyway, take another look and see what you think:
- CodePen: https://codepen.io/svinkle/pen/pjVepz/
- GitHub repo: https://github.com/svinkle/launchy

/cc @davatron5000 @joe-watkins @mpiotrowicz @jeryj @scottaohara
